Refactoring

zhss1983 opened the pull request:
Минимум изменений, никакого рефакторинга не работающих фрагментов и импортов.
- Вынести все константы из проекта в отдельную папку.
- В дополнительном разбиении по файлам в зависимости от назначения нужды не вижу, итак уже разделены.
- В именах констант использовать постфикс отображающий где используется константа. 
- Значение констант равно самим себе.
- Объединил файлы start_conversation и bot_menu
- Вынести все jobs из файла bots в новый отдельный файл jobs

zhss1983 committed:
Changed the names of constants and their values where necessary.

zhss1983 committed:
Combined the bot_menu and start_conversation files. Replaced the "start" command with a constant that was previously created, but was not used in the project.

zhss1983 committed:
Highlighted unused constants.

zhss1983 committed:
I moved the constants from the project to a separate folder. I didn't make any changes, the status can be viewed by the previous commit.

veselij commented on constants/command_constants.py:
```diff
@@ -0,0 +1,19 @@
+OVERDUE_REQUESTS_COMMAND = "overdue_requests_command"
+
+REPEAT_COMMAND = "repeat_after_one_hour_command"
+SEND_SOON_COMMAND = "send_soon_command"  # Не используемая константа
+ALREADY_SEND_COMMAND = "already_send_command"  # Не используемая константа
```
а зачем нам неиспользуемые? удалит их тогда надо

zhss1983 replied:
Я не знаю кто и для чего их создавал, может быть это часть будущего кода. Так можно много чего подчистить.
Эти я удалю, но часто были случаи когда какие-то изменения ложились на уже замердженые и в последствии другими людьми изменённые, появлялись ошибки. Так было у нас когда в команде работали.

veselij commented on src/start_conversation.py:
```diff
@@ -3,6 +3,7 @@
 
 from constants import command_constants as cmd_const
 from constants import states
+from core.config import URL_SERVICE_RULES
```
мне кажется вот это констант, а не конфиг параметр, поэтому должна лежать в константах, @AntonZelinsky  @MatMerd  как думаете?

zhss1983 replied:
Согласен, недосмотрел.

AntonZelinsky replied:
Сложно, с одной стороны я согласен с вами, а с другой стороны, чтобы изменить ссылку, нужно менять код приложения. Если она в конфиге будет лежать, то достаточно изменить инвайрмент переменную и перезапустить деплой — этого будет достаточно.
А даже если вы решили вынести в константы, то их нужно в обычный файл констант, а не создавать отдельный файл для одной константы constants/url_paths.py. 
И у меня такой же вопрос, как выше уже писал Саша: `а зачем вынеc папку констант из src?`?

zhss1983 replied:
То есть лучше в config щставить, но сделать так что бы подгружалось из .env?

zhss1983 replied:
а зачем вынеc папку констант из src? - В задании чёрным по белому написано: Вынести все константы из проекта в отдельную папку. Я вынес. Судя по вопросам не верно понял? Судя по всему формулировка имелась ввиду "Перенести все константы проекта находящиеся не в файлах констант в файлы с константами."

veselij replied:
Да, русский язык подвёл. Имелось ввиду просто все константы проекта. Верни пожалуйста под src

zhss1983 replied:
Выполнил.

AntonZelinsky replied:
Это более правильный вариант.

veselij commented on constants/command_constants.py:
```diff
@@ -0,0 +1,19 @@
+SEND_SOON_COMMAND = "send_soon_command"  # Не используемая константа
+ALREADY_SEND_COMMAND = "already_send_command"  # Не используемая константа
+
+IS_EXPERT_COMMAND = "is_expert_command"
+NOT_EXPERT_COMMAND = "not_expert_command"
+REGISTR_AS_EXPERT_COMMAND = "registr_as_expert_command"
+SUPPORT_OR_CONSULT_COMMAND = "support_or_consult_command"
+AFTER_REGISTR_MESSAGE_COMMAND = "after_registr_message_command"
+TIMEZONE_COMMAND = "timezone_command"
+SKIP_TIMEZONE_COMMAND = "skip_timezone_command"
```
это же по факту не команды, а callback_data для диалогов. может т вынесем в отдельный файл callback_data.py @AntonZelinsky @MatMerd, команд то всего две "start" и "menu"?

zhss1983 replied:
Сделал.
Вообще есть ещё "cansel", но её в константы вносить не стали. Я добавил по аналогии.

AntonZelinsky replied:
`cancel` нужно вообще удалить. Это плохой копипаст из примеров.

zhss1983 replied:
Он используется в start_conversation, параметр fallbacks=[cancel_command_handler]. Если удаляем, то он за собой тянет и соответствующий метод: async def cancel(update: Update, context: ContextTypes.DEFAULT_TYPE)

zhss1983 replied:
Ставлю fallbacks=[]
Далее выпиливаю всё связанное с cansel.
Бот запускается.

zhss1983 replied:
Ещё не плохо было бы переименовать command_constants просто в commands. Что скажете?

zhss1983 replied:
По всей видимости это не требуется.

veselij reviewed (changes requested):
а зачем вынеc  папку констант из src?

zhss1983 committed:
Deleted unused constants

zhss1983 committed:
Moved callbacks to callback_data file

zhss1983 committed:
Add cansel command to constants

zhss1983 committed:
Restored constants folder and moved URL_SERVICE_RULES to enviroments

zhss1983 committed:
Removed cansel_comand and it's methods.

zhss1983 committed:
Removed cansel_comand and it's methods. Fix PEP8 error

AntonZelinsky commented on src/constants/command_constants.py:
```diff
@@ -1,19 +1,2 @@
-TIMEZONE_COMMAND = "timezone_command"
-SKIP_TIMEZONE_COMMAND = "skip_timezone_command"
+START_COMMAND = "start"
+MENU_COMMAND = "menu"
```
У нас будут всего две команды и для их нет смысла делать отдельный файл с константами. Если очень хочется, то можно сделать константу на уровне файл в котором они используются, но и от этого нет смысла. Ищем баланс в гибкости, целесообразности,  поддерживаемости и удобстве использования.

zhss1983 replied:
Дак может ну их вовсе, оставить просто текст в коде? Он по всему применятся будет только в одном месте?

zhss1983 replied:
Удалил, вместо констант прописал их значения.

AntonZelinsky commented on src/constants/callback_data.py:
```diff
@@ -0,0 +1,13 @@
+OVERDUE_REQUESTS_CALLBACK = "overdue_requests_command"
+REPEAT_CALLBACK = "repeat_after_one_hour_command"
+IS_EXPERT_CALLBACK = "is_expert_command"
+NOT_EXPERT_CALLBACK = "not_expert_command"
```
Подожди, тут колбэки или команды? Из содержимого видно, что это команда. А если в отладчике увидишь переменную которую передали куда-то в другое место то никогда и не догадаешься что там находится всё ровно и наоборот.
В обычной жизни с этим не сталкнёшься, но если отлавливаешь баг, то это будет полный крах. Я однажды дебажил эту библиотеку часа 4, пытаясь понять почему работает не так как должно, и правильные имена и содержимое очень помогали это сделать. Если бы не они, то наверно пришлось бы всё переделывать на правильные имена и заняло бы поиск проблемы и решения не 4 часа, а намного больше..

Эти штуки используются в инлайновых кнопках и самих колбэк хэндлерах:
```
NOT_EXPERT_CALLBACK = "not_expert_command"
InlineKeyboardButton("Нет", callback_data=callback.NOT_EXPERT_CALLBACK),
CallbackQueryHandler(not_expert_callback, pattern=callback.NOT_EXPERT_CALLBACK),
```
Для кнопки он нужен для того, чтобы передать какую-то информацию нашей системе, куда происходит переход. То есть, по сути это команда, для того чтобы сделать этот переход (если представлять, что была нажата какая-то кнопка), я со скрипом согласен на паттерн или колбэк_паттерн, но никак ни на колбэк. К тому же, у тебя везде одинаковые названия, у функций и констант. 

колбэк — это всегда метод вызываемый при каком-то событии.

zhss1983 replied:
То что в значении применяется постфикс command, а в названии постфикс callback - это мой косяк. Я переименовал названия констант в соответствии с файлом в который и поместил, но забыл про сами значения.
Теперь непосредственно с именами. Я понимаю что такое колбэк, но не понимаю какие имена требуются для констант. В моём понимании если в качестве параметра требуется callback, то и передавать надо callback. И названия у констант и сами они должны быть соответствующие.

zhss1983 replied:
"К тому же, у тебя везде одинаковые названия, у функций и констант." А в чём тут проблема?

zhss1983 replied:
Последние изменения корректны? Или я опять всё не верно понял?

AntonZelinsky replied:
Всё окей. Сергей, добавь, пожалуйста, в файлик с соглашениями и хорошими практиками информацию о том, как нужно именовать переменные в боте.
Например:

- `menu_command_handler = CommandHandler("menu", menu)` — для команд хэндлеров добавляем постфикс `_command_handler`
- `menu_callback_handler = CallbackQueryHandler(handling_menu_button_click_callback)` — `_callback_handler`
- `InlineKeyboardButton("Да", callback_data=callback.CALLBACK_REGISTR_AS_EXPERT_COMMAND),` для колбэк даты ...
- состояния сюда же.
Про то где и в каком файле должно быть. Это позволит всем придерживаться одного стиля

AntonZelinsky commented on src/start_conversation.py:
```diff
@@ -110,15 +112,33 @@ async def timezone_message_callback(update: Update, context: ContextTypes.DEFAUL
-start_command_handler = CommandHandler("start", start)
-cancel_command_handler = CommandHandler("cancel", cancel)
+
+start_command_handler = CommandHandler(cmd_const.START_COMMAND, start)
+menu_command_handler = CommandHandler(cmd_const.MENU_COMMAND, callback=menu)
```
Если эти штуки нигде не переиспользуются, то нет смысла выносить их в отдельные переменные, пусть как CallbackQueryHandler создаются прямо в структуре ConversationHandler

zhss1983 replied:
Выше вопрос задавал. Снимаю.

zhss1983 replied:
Сделано.

AntonZelinsky commented on src/start_conversation.py:
```diff
@@ -127,20 +147,22 @@ async def cancel(update: Update, context: ContextTypes.DEFAULT_TYPE):
+    fallbacks=[],
 )
+
+callback_menu_handler = CallbackQueryHandler(handling_menu_button_click_callback)
```
Не нашёл где это используется

zhss1983 replied:
А оно и не используется. Это фрагмент из файла bot_menu.py Он предполагает вызов меню. Я даже добавлял его в код так что бы меню вызывалось, но тот код вы забраковали. Оставил как есть. Для чего нужен будет или для чего применялся надо наверное у Андрея спрашивать.

zhss1983 replied:
Изменил код таким образом что бы меню можно было вызвать по команде. Изменения минимальны.

zhss1983 replied:
Теперь меню вызывается по команде /menu. Я не автор этой части, старался минимизировать вмешательство в чужой код.

veselij replied:
там толькло теперь надо сделать чтобы меню переводило в стейт меню, чтобы callback  нажатия кнопок меню отрабатывал. Но это уже вторая таска рефакторинга стейтов плавно началась. давай в отдельном PR уже

zhss1983 replied:
Ок.

zhss1983 committed:
Rename callback constants to command ones.

zhss1983 committed:
Fix REPEAT_COMMAND constant

zhss1983 committed:
Fix REPEAT_COMMAND constant

zhss1983 committed:
Изменил имена констант на CALLBACK_<имя метода который он вызывает>_COMMAND для того чтобы сразу становилось ясно для чего нужна та или иная константа.
Перенёс запуск меню в start_conversation по команде /menu.

zhss1983 committed:
I changed the names of the constants to CALLBACK_<the name of the method that it calls>_COMMAND so that it immediately becomes clear what this or that constant is needed for.
Moved the menu launch to start_conversation by the /menu command.
I saved callback_menu_handler in case it is imported in the future, the code is not mine, I do not know how it will be reused.
